audioipc/src/ipccore.rs

```diff
@@ -0,0 +1,717 @@
+// Copyright © 2021 Mozilla Foundation
+//
+// This program is made available under an ISC-style license.  See the
+// accompanying file LICENSE for details
+
+use std::io::Result;
+use std::sync::{mpsc, Arc};
+use std::thread;
+
+use mio::{event::Event, Events, Interest, Poll, Registry, Token, Waker};
+use slab::Slab;
+
+use crate::messages::AssocRawPlatformHandle;
+use crate::rpccore::{make_client, make_server, Client, Handler, Proxy, Server};
+use crate::{
+    codec::Codec,
+    codec::LengthDelimitedCodec,
+    sys::{self, RecvMsg, SendMsg},
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+use std::fmt::Debug;
+
+#[cfg(unix)]
+use crate::cmsg;
+#[cfg(windows)]
+use crate::duplicate_platform_handle;
+
+const WAKE_TOKEN: Token = Token(!0);
+
+// Requests sent by an EventLoopHandle to be serviced by
+// the handle's associated EventLoop.
+enum Request {
+    // See EventLoop::add_connection
+    AddConnection(
+        sys::Pipe,
+        Box<dyn Driver + Send>,
+        mpsc::Sender<Result<Token>>,
+    ),
+    // See EventLoop::shutdown
+    Shutdown,
+    // See EventLoop::wake_connection
+    WakeConnection(Token),
+}
+
+// EventLoopHandle is a cloneable external reference
+// to a running EventLoop, allowing registration of
+// new client and server connections, in addition to
+// requesting the EventLoop shut down cleanly.
+#[derive(Clone, Debug)]
+pub struct EventLoopHandle {
+    waker: Arc<Waker>,
+    requests_tx: mpsc::Sender<Request>,
+}
+
+impl EventLoopHandle {
+    pub fn bind_client<C: Client + 'static>(
+        &self,
+        client_pipe: sys::Pipe,
+    ) -> Result<Proxy<<C as Client>::Request, <C as Client>::Response>>
+    where
+        <C as Client>::Request: Serialize + Debug + AssocRawPlatformHandle + Send,
+        <C as Client>::Response: DeserializeOwned + Debug + AssocRawPlatformHandle + Send,
+    {
+        let (handler, mut proxy) = make_client::<C>();
+        let driver = Box::new(FramedDriver::new(handler));
+        let token = self.add_connection(client_pipe, driver)?;
+        proxy.connect_event_loop(self.clone(), token);
+        Ok(proxy)
+    }
+
+    pub fn bind_server<S: Server + Send + 'static>(
+        &self,
+        server: S,
+        server_pipe: sys::Pipe,
+    ) -> Result<()>
+    where
+        <S as Server>::Request: DeserializeOwned + Debug + AssocRawPlatformHandle + Send,
+        <S as Server>::Response: Serialize + Debug + AssocRawPlatformHandle + Send,
+    {
+        let handler = make_server::<S>(server);
+        let driver = Box::new(FramedDriver::new(handler));
+        self.add_connection(server_pipe, driver)?;
+        Ok(())
```

Comment:
nit: I find using `Ok(self.add_connection(server_pipe, driver)?)` and adding a `trace!(..)` printing the token here helpful for me to understand the code better. But I don't have a strong opinion.

Reply:
Done.